rust/src/producer.rs

```diff
@@ -93,23 +116,80 @@ impl Producer {
         };
         let logger = log::logger(option.logging_format());
         let settings = build_producer_settings(&option, &client_option);
-        let mut client = Client::new(&logger, client_option, settings)?;
-        client.set_transaction_checker(transaction_checker);
+        let client = Client::new(&logger, client_option, settings)?;
+        let option = Arc::new(RwLock::new(option));
         Ok(Producer {
             option,
             logger,
             client,
+            transaction_checker: Some(transaction_checker),
+            shutdown_tx: None,
         })
     }
 
+    async fn get_resource_namespace(&self) -> String {
+        let option_guard = self.option.read();
+        let resource_namespace = option_guard.await.namespace().to_string();
+        resource_namespace
+    }
+
     /// Start the producer
     pub async fn start(&mut self) -> Result<(), ClientError> {
-        self.client.start().await?;
-        if let Some(topics) = self.option.topics() {
+        let (telemetry_command_tx, mut telemetry_command_rx) = mpsc::channel(16);
+        let telemetry_command_tx: mpsc::Sender<pb::telemetry_command::Command> =
+            telemetry_command_tx;
+        self.client.start(telemetry_command_tx).await?;
+        let option_guard = self.option.read().await;
+        let topics = option_guard.topics();
+        if let Some(topics) = topics {
             for topic in topics {
                 self.client.topic_route(topic, true).await?;
             }
         }
+        drop(option_guard);
+        let transaction_checker = self.transaction_checker.take();
+        if transaction_checker.is_some() {
+            self.transaction_checker = Some(Box::new(|_, _| TransactionResolution::UNKNOWN));
```

Comment:
Why not simply take the transaction checker and leave None afterward?

Reply:
> Why not simply take the transaction checker and leave None afterward?

In `send_transaction_message` function, the producer uses `has_transaction_checker` function to check if it's legal to call this function. If we leave None, we have no idea whether it's a transaction producer.

Reply:
You are right.